### PR TITLE
Adds a cumulative table to the HOC Signups dashboard as well as HOC 2014 data.

### PR DESCRIPTION
**Pull request by ashercodeorg**

### CODE REVIEW

File: dashboard/app/controllers/admin_reports_controller.rb [hunk 1]
```diff
@@ -206,10 +206,50 @@ def hoc_signups
...
+        all.
+        map{|row| [row[:created_at_day].to_s, row[:count].to_i]}
+
+    # Construct the hash {MM-DD => [count2014, count2015]}.
+    # Start by constructing the key space as the union of the MM-DD dates for
+    # data_2014 and data_2015.
+    require 'set'
+    dates = SortedSet.new []
+    data_2014.each do |day|
+      dates.add(day[0][5..11])
```
Comment (Hamms) [inline dashboard/app/controllers/admin_reports_controller.rb]: why `5..11`? A comment explaining why this range was chosen would be nice

File: dashboard/app/controllers/admin_reports_controller.rb [hunk 2]
```diff
@@ -206,10 +206,50 @@ def hoc_signups
...
+    end
+    # Then populate the keys of our hash {date=>[count2014,count2015], ..., date=>[...]} with dates.
+    data_by_day = {}
+    dates.each do |date|
+      data_by_day[date] = [0, 0]
+    end
+    # Finally populate the values of our hash.
+    data_2014.each do |day|
+      data_by_day[day[0][5..11]] = [day[1], 0]
```
Comment (Hamms) [inline dashboard/app/controllers/admin_reports_controller.rb]: given that we can be confident that `data_by_day[date]` is already initialized for all dates, we could here just do `data_by_day[day[0][5..11]][0] = day[1]`, and below `data_by_day[day[0][5..11]][1] = day[1]`

Comment (Hamms) [inline dashboard/app/controllers/admin_reports_controller.rb]: Alternatively, something like

``` ruby
data_by_day[date] = {2014 => 0, 2015 => 0}
data_by_day[day[0][5..11]][2014] = day[1]
```

Would make the whole thing significantly less implicit

File: dashboard/app/controllers/admin_reports_controller.rb [hunk 3]
```diff
@@ -206,10 +206,50 @@ def hoc_signups
...
+        group_and_count(Sequel.as(Sequel.qualify(:forms, :created_at).cast(:date),:created_at_day)).
+        order(:created_at_day).
+        all.
+        map{|row| [row[:created_at_day].to_s, row[:count].to_i]}
+
+    # Construct the hash {MM-DD => [count2014, count2015]}.
+    # Start by constructing the key space as the union of the MM-DD dates for
+    # data_2014 and data_2015.
+    require 'set'
+    dates = SortedSet.new []
```
Comment (Hamms) [inline dashboard/app/controllers/admin_reports_controller.rb]: is there a reason this is a `SortedSet`?

File: dashboard/app/controllers/admin_reports_controller.rb (hunk 2 above)
Comment (ashercodeorg) [inline dashboard/app/controllers/admin_reports_controller.rb]: Done (the former).

File: dashboard/app/controllers/admin_reports_controller.rb (hunk 1 above)
Comment (ashercodeorg) [inline dashboard/app/controllers/admin_reports_controller.rb]: Changed to 5..9, now with a comment.

File: dashboard/app/controllers/admin_reports_controller.rb (hunk 3 above)
Comment (ashercodeorg) [inline dashboard/app/controllers/admin_reports_controller.rb]: There isn't (though there was). Changed to Set.

File: dashboard/app/controllers/admin_reports_controller.rb [hunk 4]
```diff
@@ -206,10 +206,51 @@ def hoc_signups
...
+        all.
+        map{|row| [row[:created_at_day].to_s, row[:count].to_i]}
+
+    # Construct the hash {MM-DD => [count2014, count2015]}.
+    # Start by constructing the key space as the union of the MM-DD dates for
+    # data_2014 and data_2015.
+    require 'set'
+    dates = Set.new []
+    data_2014.each do |day|
+      # Here and below, we use [5..9] to extract 'MM-DD' from 'YYYY-MM-DD'.
```
Comment (Hamms) [inline dashboard/app/controllers/admin_reports_controller.rb]: Could we do that above?

``` ruby
map{|row| [row[:created_at_day].strftime("%m-%d"), row[:count].to_i]}
```

would be delightfully clear

Comment (ashercodeorg) [inline dashboard/app/controllers/admin_reports_controller.rb]: Done.

File: dashboard/app/views/admin_reports/hoc_signups.html.haml [hunk 5]
```diff
@@ -1,39 +1,70 @@
...
-      data.addColumn('string', 'Date');
-      data.addColumn('number', 'Count');
-      data.addRows(signups_by_day);
+    var data_by_day = #{data_by_day.try(:to_json)};
+
+    if (data_by_day.length > 0) {
+      // Generate the signups and cumulative breakdowns from the data, if data exists.
+      signups_by_day = [[data_by_day[0][0], data_by_day[0][1][0], data_by_day[0][1][1]]];
+      cumulatives_by_day = [[data_by_day[0][0], data_by_day[0][1][0], data_by_day[0][1][1]]];
+      for (var i = 1; i < data_by_day.length; i++) {
```
Comment (Hamms) [inline dashboard/app/views/admin_reports/hoc_signups.html.haml:44]: why are we beginning counting from 1?

Comment (ashercodeorg) [inline dashboard/app/views/admin_reports/hoc_signups.html.haml:44]: So that we can reference cumulatives[prev_day] when computing cumulatives[next_day].

**Comment by ashercodeorg** [comment]
Merging, despite the hound violation, since I can't reproduce it locally running lint.

**Pull request merged**